### ISSUE TITLE
Move to using a SidreDataCollection for state variable storage

### ISSUE THREAD

**Issue by jamiebramwell** [LLNL/serac#29]
Look at [NamedFieldsMap](https://github.com/mfem/mfem/blob/b584a1a82dc17a0e46b5e3c760f8edcb3107ab8d/fem/datacollection.hpp#L29) for guidance on named lookups to the data collection.

**Comment by joshessman-llnl** [comment]
## MFEMSidreDataCollection

The `MFEMSidreDataCollection` class is used to store the state of a simulation, namely, the mesh, associated fields, and simulation information like the current time and frame number.
The state is stored in a global Sidre tree (following the Conduit Blueprint structure) and can be saved in HDF5 (and JSON) format for visualization or restart purposes.

An abbreviated usage looks like the following:

```cpp
// Finite Element setup
mfem::Mesh mesh = loadMeshFromFile(mesh_filename);
mfem::H1_FEC fec(1,2);
mfem::FES fes(mesh, fec);
mfem::GridFunction soln(fes);

// Datastore Setup
MFEMSidreDataCollection dc;
dc.SetMesh(mesh);
dc.RegisterField("solution", soln);
dc.SetTime(0);

// for each iteration
dc.SetTime(t);
dc.Save();
```

## Restart

When initialized with a mesh, the `MFEMSidreDataCollection` extracts and retains information in the datastore sufficient to reconstruct a mesh.
It also contains data sufficient to reconstruct `mfem::GridFunctions`.  However, this is not implemented, so a newly constructed `MFEMSidreDataCollection` cannot simply
`Load()` from a file and have a state identical to the `MFEMSidreDataCollection` that was `Save()`d.  

For this reason, data is externally reconstructed in MARBL (pseudocode):

```cpp
// Finite Element setup
mfem::Mesh mesh = loadMeshFromFile(mesh_filename);
mfem::H1_FEC fec(1,2);
mfem::FES fes(mesh, fec);
mfem::GridFunction soln(fes);

// The data is "manually" pulled out of the datastore
// and used to repopulate the gridfunction 
DataStore ds = loadDSFromFile(datastore_filename);
FillFieldFromDataStore(ds, soln); 

MFEMSidreDataCollection dc(ds);

// mfem::Mesh& mesh = dc.GetMesh(); // Returns nullptr, mesh was not reconstructed
// Returns nullptr, field was not reconstructed even though its data is in the Datacollection
// mfem::GridFunction& soln = dc.GetField("solution");

dc.SetMesh(mesh);
dc.RegisterField("solution", soln);
dc.SetTime(0); // Data is actually from the restart point/checkpoint, but it's reset to zero
```

Note that the state is completely owned by the user of the datastore - the datastore may own some of the
underlying data, but the actual `mfem::Mesh` and `mfem::GridFunction` objects are owned by the user.

## True "Load" functionality

Alternatively, the `Load()` function could be modified to be the true inverse of `Save()` such that state is completely restored:

```cpp
MFEMSidreDataCollection dc;
dc.Load(datastore_filename);

mfem::Mesh& mesh = dc.GetMesh(); // Mesh has already been configured
mfem::GridFunction& soln = dc.GetField("solution"); // Field already exists and is populated
double t = dc.GetTime(); // Will be the last checkpoint, not zero
```

This is useful because the from-scratch construction of state objects is not actually required in a restart situation - the datastore
already has all the information it needs.  It also avoids the using twice as much memory to load and copy the data, as the objects
are constructed within the datastore directly using the datastore's memory.

That being said, in this case the ownership is dependent on whether the current simulation is a restart - in restarts, the `MFEMSidreDataCollection` would own
the state, whereas in a nominal run it would be owned by the user.  This could present some complications to user code as initialization looks different
between nominal and restart runs.

## Implementation within Serac

In Serac, the relevant logic is mostly confined to the `FiniteElementState` class which stores a mesh, FEColl, FESpace, GridFunction, etc.  It owns all its members exclusively (with the exception of the mesh, which is shared) - although FESpace and FEColl can be accessed through a GridFunction, they are not owned by the GridFunction, which is one of the reasons that the class exists - to encapsule a GridFunction and the subobjects it is composed of.

Potentially this class could be modified to support the wrapping
of references to simulation state in addition to its current wrapping of ownership of state, e.g., with a `MaybeOwner<T>` helper class that wraps `std::variant<std::reference_wrapper<T>, std::unique_ptr<T>>`.

This might require modifications to all physics module constructors to handle restarts, as they all currently construct a new `FiniteElementState`.

To avoid requiring the physics modules to implement restart logic, the `FiniteElementState` constructor (which would have a reference to the `DataCollection`) could ignore the other options provided by the module (FEColl type, ordering, vector dimension, etc) if the datastore is not empty:

```cpp
FiniteElementState(mfem::ParMesh& mesh, axom::sidre::MFEMSidreDataCollection& dc, FEStateOptions&& options)
{
  if (dc.HasField(options.name))
  {
    // It's a restart
    mesh_ = dc.GetMesh();
    gf_ = dc.GetField(options.name); // Owned by the DC
    fespace_ = gf_.FESpace(); // Owned by the DC
    coll_ = fespace_.FEColl(); // Owned by the DC
  }

  else
  {
    // Nominal run
    mesh_ = mesh;
    coll_ = /* construct from scratch ... */;
    fespace_ = /* construct from scratch ... */;
    gf_ = make_unique<mfem::GridFunction>(fespace_);
  }
}
```

## TLDR

The primary concern with the "true" reload functionality where the state of the datastore is perfectly restored, as opposed to unconditional full initialization followed by optional overwriting during restarts is the need to modify constructors.  Since state initialization is already mostly confined to one class, this could be minimal.

**Comment by joshessman-llnl** [comment]
## MFEMSidreDataCollection

The `MFEMSidreDataCollection` class is used to store the state of a simulation, namely, the mesh, associated fields, and simulation information like the current time and frame number.
The state is stored in a global Sidre tree (following the Conduit Blueprint structure) and can be saved in HDF5 (and JSON) format for visualization or restart purposes.

An abbreviated usage looks like the following:

```cpp
// Finite Element setup - init()
{
  mfem::Mesh mesh = loadMeshFromFile(mesh_filename);
  mfem::H1_FEC fec(1,2);
  mfem::FES fes(mesh, fec);
  mfem::GridFunction soln(fes);
}

// Datastore Setup
MFEMSidreDataCollection dc;
dc.SetMesh(mesh);
dc.RegisterField("solution", soln);
dc.SetTime(0);

// for each iteration
dc.SetTime(t);
dc.Save();
```

## Restart

When initialized with a mesh, the `MFEMSidreDataCollection` extracts and retains information in the datastore sufficient to reconstruct a mesh.
It also contains data sufficient to reconstruct `mfem::GridFunctions`.  However, this is not implemented, so a newly constructed `MFEMSidreDataCollection` cannot simply
`Load()` from a file and have a state identical to the `MFEMSidreDataCollection` that was `Save()`d.  

For this reason, the only option is to construct data externally (pseudocode):

```cpp
// Finite Element setup - init()
{
  mfem::Mesh mesh = loadMeshFromFile(mesh_filename);
  mfem::H1_FEC fec(1,2);
  mfem::FES fes(mesh, fec);
  mfem::GridFunction soln(fes);
}

// Optional restart - restart()
{
  // The data is "manually" pulled out of the datastore
  // and used to repopulate the gridfunction 
  DataStore ds = loadDSFromFile(datastore_filename);
  FillFieldFromDataStore(ds, soln); 
}

MFEMSidreDataCollection dc(ds);

// mfem::Mesh& mesh = dc.GetMesh(); // Returns nullptr, mesh was not reconstructed
// Returns nullptr, field was not reconstructed even though its data is in the Datacollection
// mfem::GridFunction& soln = dc.GetField("solution");

dc.SetMesh(mesh);
dc.RegisterField("solution", soln);
dc.SetTime(0); // Data is actually from the restart point/checkpoint, but it's reset to zero
```

Note that the state is completely owned by the user of the datastore - the datastore may own some of the
underlying data, but the actual `mfem::Mesh` and `mfem::GridFunction` objects are owned by the user.

## True "Load" functionality

Alternatively, the `Load()` function could be modified to be the true inverse of `Save()` such that state is completely restored:

```cpp
MFEMSidreDataCollection dc;
dc.Load(datastore_filename);

mfem::Mesh& mesh = dc.GetMesh(); // Mesh has already been configured
mfem::GridFunction& soln = dc.GetField("solution"); // Field already exists and is populated
double t = dc.GetTime(); // Will be the last checkpoint, not zero
```

This is useful because the from-scratch construction of state objects is not actually required in a restart situation - the datastore
already has all the information it needs.  It also avoids the using twice as much memory to load and copy the data, as the objects
are constructed within the datastore directly using the datastore's memory.

That being said, in this case the ownership is dependent on whether the current simulation is a restart - in restarts, the `MFEMSidreDataCollection` would own
the state, whereas in a nominal run it would be owned by the user.  This could present some complications to user code as initialization looks different
between nominal and restart runs.

## Implementation within Serac

In Serac, the relevant logic is mostly confined to the `FiniteElementState` class which stores a mesh, FEColl, FESpace, GridFunction, etc.  It owns all its members exclusively (with the exception of the mesh, which is shared) - although FESpace and FEColl can be accessed through a GridFunction, they are not owned by the GridFunction, which is one of the reasons that the class exists - to encapsule a GridFunction and the subobjects it is composed of.

Potentially this class could be modified to support the wrapping
of references to simulation state in addition to its current wrapping of ownership of state, e.g., with a `MaybeOwner<T>` helper class that wraps `std::variant<std::reference_wrapper<T>, std::unique_ptr<T>>`.

This might require modifications to all physics module constructors to handle restarts, as they all currently construct a new `FiniteElementState`.

To avoid requiring the physics modules to implement restart logic, the `FiniteElementState` constructor (which would have a reference to the `DataCollection`) could ignore the other options provided by the module (FEColl type, ordering, vector dimension, etc) if the datastore is not empty:

```cpp
FiniteElementState(mfem::ParMesh& mesh, axom::sidre::MFEMSidreDataCollection& dc, FEStateOptions&& options)
{
  if (dc.HasField(options.name))
  {
    // It's a restart
    mesh_ = dc.GetMesh();
    gf_ = dc.GetField(options.name); // Owned by the DC
    fespace_ = gf_.FESpace(); // Owned by the DC
    coll_ = fespace_.FEColl(); // Owned by the DC
  }

  else
  {
    // Nominal run
    mesh_ = mesh;
    coll_ = /* construct from scratch ... */;
    fespace_ = /* construct from scratch ... */;
    gf_ = make_unique<mfem::GridFunction>(fespace_);
  }
}
```

## TLDR

The primary concern with the "true" reload functionality where the state of the datastore is perfectly restored, as opposed to unconditional full initialization followed by optional overwriting during restarts is the need to modify constructors.  Since state initialization is already mostly confined to one class, this could be minimal.
